APMrover2/mode_breadcrumb.cpp

```diff
@@ -0,0 +1,183 @@
+#include "mode.h"
+#include "Rover.h"
+
+
+bool ModeBreadcrumb::_enter()
+{
+    //don't call enter again if _crumbs exists
+    if (_crumbs != nullptr) {
+        return false;
+    }
+
+    if (_crumbs_sem == nullptr){
+        // create semaphore
+        _crumbs_sem = hal.util->new_semaphore();
+        if (_crumbs_sem == nullptr) {
+            return false;
+        }
+    }
+
+    _current_crumb = nullptr;
+
+    // allocate crumbs queue
+    _crumbs = (Location*)calloc(crumbs_max, sizeof(Location));
+
+    _crumbs_tail = 0;
+    _crumbs_head = 0;
+
+    lonely_mode = nullptr;
+
+    //reuse guided mode
+    if (!ModeGuided::_enter()) {
+        free(_crumbs);
+        return false;
+    }
+
+    return true;
+}
+
+void ModeBreadcrumb::_exit()
+{
+    //reset lonely mode in case we go back into breadcrumb mode
+    lonely_mode = nullptr;
+
+    _current_crumb = nullptr;
+
+    if (_crumbs != nullptr) {
+        free(_crumbs);
+        //stop usage of freed data
+        _crumbs = nullptr;
+    }
+}
+
+void ModeBreadcrumb::run_lonely_mode()
+{
+    if (lonely_mode == nullptr) {
+        rover.mode_hold.enter();
+        lonely_mode = &rover.mode_hold;
+
+        gcs().send_text(MAV_SEVERITY_INFO, "Breadcrumb: Lonely; %s", lonely_mode->name4());
+    }
+
+    lonely_mode->update();
+}
+
+void ModeBreadcrumb::update()
+{
+
+
+    float distance_to_breadcrumb = ModeGuided::get_distance_to_destination();
+
+    //we have or are going to a breadcrumb
+    //get_distance_to_destination returns exactly 0.0f when destination is reached
+    if (_current_crumb != nullptr && distance_to_breadcrumb != 0.0f) {
+
+        Vector3f to_vehicle = location_3d_diff_NED(rover.current_loc, target_loc);
```

Comment:
Did you mean to take the unit vector from this vector before doing more maths on it?

Reply:
Ah, forgot that. 
I also noticed that whenever the Rover races towards the copter and then the copter makes a turn, the Rover stops and just circles. This was in my SITL demo. I thought it might be due to the closure speed being too high but dropping that didn't solve it. I tried investigating it for a while, but I'm not really sure why this is happening. Is it maybe because I don't have loop detection?